### ISSUE TITLE
ImplementedReturnTypeMismatch on template trait abstract method

### ISSUE THREAD

**Issue by ptomulik** [vimeo/psalm#6937]
Two template traits used by a class, one trait declares an abstract method, other implements it. When used, the ImplementedReturnTypeMismatch error is reported.

https://psalm.dev/r/07c75f40bc

**Comment by weirdan** [comment]
Simplified and condensed a bit: https://psalm.dev/r/d6d08bda2f

**Comment by ptomulik** [comment]
I'd like to try to fix this, where shall I start?

**Comment by weirdan** [comment]
I usually start with running Psalm on the failing snippet with `--debug-emitted-issues --threads=1 --no-cache` flags. It produces backtrace pinpointing the place where the error is emitted. Then work back from there.

`PSALM_ALLOW_XDEBUG=1 psalm --no-cache --threads=1` allows step debugging with Xdebug.

**Comment by ptomulik** [comment]
It looks like [MethodComparator::create](https://github.com/vimeo/psalm/blob/6097e027730373dcc6473fbf334484ad904e37b9/src/Psalm/Internal/Analyzer/MethodComparator.php#L41) gets called twice for the above example. It first compares the two traits, then it attempts to compare ``Test`` class with ``AnotherTrait`` and the test fails. I've added some diagnostic messages and got:

```
comparing: 'AggregateTrait' with 'AnotherTrait': success                        
comparing: 'Test' with 'AnotherTrait': failure                                  
```

I observed that the comparison is successfull as the implementer class(like) is a trait. This may also be infered from the code of  [MethodComparator::compareMethodDocblockReturnType()](https://github.com/vimeo/psalm/blob/6097e027730373dcc6473fbf334484ad904e37b9/src/Psalm/Internal/Analyzer/MethodComparator.php#L657). If its not a trait, then (``Test`` class in our case), then the [code](https://github.com/vimeo/psalm/blob/6097e027730373dcc6473fbf334484ad904e37b9/src/Psalm/Internal/Analyzer/MethodComparator.php#L696) that transforms template parameters is not executed, and the information about "int" type provided as parameter to ``@template-use AnotherTrait`` is not used.

The question is now, where to dig further to find an solution?
